File: python-deprecated/utils/image_storage.py
"""S3-backed storage helpers for uploaded metadata images."""

import logging
from abc import ABC, abstractmethod
from datetime import UTC, datetime
from uuid import uuid4

import aioboto3

from db.config import settings

logger = logging.getLogger(__name__)


class ImageStorage(ABC):
    """Abstract storage contract for uploaded images."""

    @abstractmethod
    async def store_image(self, key: str, content: bytes, content_type: str) -> str:
        """Store image bytes and return storage key."""

    @abstractmethod
    async def retrieve_image(self, key: str) -> tuple[bytes, str] | None:
        """Retrieve image bytes and content type by key."""


class S3ImageStorage(ImageStorage):
    """Stores uploaded images in S3/R2-compatible object storage."""

    def __init__(self) -> None:
        if not settings.image_upload_enabled:
            raise ValueError(
                "S3 image uploads require s3_endpoint_url, s3_access_key_id, "
                "s3_secret_access_key, and s3_bucket_name to be configured."
            )

    def _get_s3_client(self):
        session = aioboto3.Session()
        return session.client(
            "s3",
            endpoint_url=settings.s3_endpoint_url,
            aws_access_key_id=settings.s3_access_key_id,
            aws_secret_access_key=settings.s3_secret_access_key,
            region_name=settings.s3_region,
        )

    async def store_image(self, key: str, content: bytes, content_type: str) -> str:
        async with self._get_s3_client() as s3:
            await s3.put_object(
                Bucket=settings.s3_bucket_name,
                Key=key,
                Body=content,
                ContentType=content_type,
                CacheControl="public, max-age=31536000, immutable",
            )
        return key

    async def retrieve_image(self, key: str) -> tuple[bytes, str] | None:
        try:
            async with self._get_s3_client() as s3:
                response = await s3.get_object(
                    Bucket=settings.s3_bucket_name,
                    Key=key,
                )
                content = await response["Body"].read()
                content_type = response.get("ContentType") or "application/octet-stream"
                return content, content_type
        except Exception:
            logger.warning("Failed to retrieve uploaded image from S3 key=%s", key)
            return None


def generate_image_storage_key(extension: str) -> str:
    """Build a safe image object key under date-based prefixes."""
    normalized_ext = extension.strip().lower().lstrip(".")
    if not normalized_ext:
        normalized_ext = "bin"

    now = datetime.now(UTC)
    return f"images/{now.strftime('%Y')}/{now.strftime('%m')}/{uuid4().hex}.{normalized_ext}"


def normalize_image_storage_key(raw_key: str) -> str:
    """Normalize and validate a user-facing storage key."""
    key = raw_key.strip().lstrip("/")
    if not key or ".." in key or "\\" in key or not key.startswith("images/"):
        raise ValueError("Invalid image key")
    return key


_storage_instance: ImageStorage | None = None


def get_image_storage() -> ImageStorage:
    """Get configured image storage backend (singleton)."""
    global _storage_instance
    if _storage_instance is None:
        _storage_instance = S3ImageStorage()
    return _storage_instance

File: backend/src/util/image_storage.rs
//! Uploaded metadata image storage (local disk or S3/R2).

use std::path::PathBuf;

use aws_sdk_s3::primitives::ByteStream;
use chrono::Utc;
use tracing::warn;
use uuid::Uuid;

use crate::config::AppConfig;

pub fn generate_image_storage_key(extension: &str) -> String {
    let normalized = extension.trim().trim_start_matches('.').to_lowercase();
    let ext = if normalized.is_empty() {
        "bin"
    } else {
        normalized.as_str()
    };
    format!(
        "images/{}/{}/{}.{}",
        Utc::now().format("%Y"),
        Utc::now().format("%m"),
        Uuid::new_v4().as_simple(),
        ext
    )
}

pub fn normalize_image_storage_key(raw_key: &str) -> Result<String, String> {
    let key = raw_key.trim().trim_start_matches('/');
    if key.is_empty() || key.contains("..") || key.contains('\\') || !key.starts_with("images/") {
        return Err("Invalid image key".into());
    }
    Ok(key.to_string())
}

pub async fn store_image(
    config: &AppConfig,
    key: &str,
    content: &[u8],
    content_type: &str,
) -> Result<String, String> {
    if config.effective_image_storage_backend() == "s3" {
        if let (Some(client), Some(bucket)) = (
            crate::util::s3_client::build_s3_client(config).await,
            crate::util::s3_client::bucket_name(config),
        ) {
            match client
                .put_object()
                .bucket(bucket)
                .key(key)
                .body(ByteStream::from(content.to_vec()))
                .content_type(content_type)
                .cache_control("public, max-age=31536000, immutable")
                .send()
                .await
            {
                Ok(_) => return Ok(key.to_string()),
                Err(e) => {
                    warn!("image_storage: S3 store failed key={key}: {e}");
                    return Err("Failed to store image in S3".into());
                }
            }
        }
        return Err("S3 image storage is not configured".into());
    }

    let dir = PathBuf::from(&config.images_dir);
    tokio::fs::create_dir_all(&dir)
        .await
        .map_err(|e| format!("Failed to create image directory: {e}"))?;
    let file_name = key.rsplit('/').next().unwrap_or(key);
    let path = dir.join(file_name);
    tokio::fs::write(&path, content)
        .await
        .map_err(|e| format!("Failed to save image: {e}"))?;
    Ok(key.to_string())
}

pub async fn retrieve_image(config: &AppConfig, key: &str) -> Option<(Vec<u8>, String)> {
    let normalized = normalize_image_storage_key(key).ok()?;

    if config.effective_image_storage_backend() == "s3"
        && let (Some(client), Some(bucket)) = (
            crate::util::s3_client::build_s3_client(config).await,
            crate::util::s3_client::bucket_name(config),
        ) {
            match client
                .get_object()
                .bucket(bucket)
                .key(&normalized)
                .send()
                .await
            {
                Ok(resp) => {
                    let content_type = resp
                        .content_type()
                        .unwrap_or("application/octet-stream")
                        .to_string();
                    if let Ok(data) = resp.body.collect().await {
                        return Some((data.into_bytes().to_vec(), content_type));
                    }
                }
                Err(e) => warn!("image_storage: S3 retrieve failed key={normalized}: {e}"),
            }
            return None;
        }

    let file_name = normalized.rsplit('/').next()?;
    let path = PathBuf::from(&config.images_dir).join(file_name);
    let bytes = tokio::fs::read(&path).await.ok()?;
    let ext = file_name.rsplit('.').next().unwrap_or("");
    let content_type = match ext {
        "png" => "image/png",
        "jpg" | "jpeg" => "image/jpeg",
        "webp" => "image/webp",
        "gif" => "image/gif",
        _ => "application/octet-stream",
    };
    Some((bytes, content_type.to_string()))
}
